XAFR - Technical Training

xafr-odoo opened the pull request:
A PR to log and facilitate the review of the completed chapters.

xafr-odoo committed:
[ADD] Chapter 1: Manifest & init

xafr-odoo committed:
[IMP] Chapter 1: Manifest & init

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/740.png)](https://mergebot.odoo.com/odoo/tutorials/pull/740)

sedi-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,13 @@
+    An application module that aims to serve as an onboarding sandbox.
+    """,
+    'application': "True",
+}
```
Don't forget to always add an empty line at the end of your files :)

xafr-odoo replied:
Ok I updated and I'll do that in the future. Why do I need to add that empty line?

sedi-odoo replied:
It is one of the PEP8 recommendations

xafr-odoo committed:
[IMP] Chapter 2: New Application

xafr-odoo committed:
[IMP] Chapter 3: Models & Basic Fields

xafr-odoo committed:
[IMP] Chapter 4: Security (+ added license in manifest to remove warning message on db launch, also removed code from estate_property.py because it was clutter that will be added in chapter 10)

sedi-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,22 @@
+    _name = "estate.property"
+    _description = "Estate properties"
+
+    name = fields.Char('Title', required=True, translate=True)
+    description = fields.Text('Description')
+    postcode = fields.Integer('Postcode', help="Your post code in 4 or 5 digits.")
+    date_availability = fields.Date('Date Availability')
+    expected_price = fields.Float('Expected Price', required=True)
+    selling_price = fields.Float('Selling Price')
+    bedrooms = fields.Integer(string='Bedrooms')
+    living_area = fields.Integer('Living Area (sqm)')
+    facades = fields.Integer('Facades')
+    garage = fields.Boolean('Garage')
+    garden = fields.Boolean('Garden')
+    garden_area = fields.Integer('Garden Area (sqm)')
+    garden_orientation = fields.Selection(
+        string='Garden Orientation',
+        selection=[('north', 'North'), ('west', 'West'), ('south', 'South'), ('east', 'East')],
+        help="The selection of the garden orientation.")
```
All good but most of the field display names are not required.

If not set, Odoo will basically deduct the display name of the field based on the field name. It'll separate it on the underscore and capitalize every word. (In case of a Many2One, Many2Many or One2Many behaviour will be slightly different but we'll see that when you'll have some). For example:

`
description -> Description
expected_price -> Expected Price
`

Basically, except `name`, `living_area` and `garden_area` none of them are required to be explicitly mentioned. Not a big issue but helps code look cleaner.

xafr-odoo committed:
[IMP] estate - Chapter 5: UI

xafr-odoo committed:
[IMP] estate - Chapter 5: UI

xafr-odoo committed:
[IMP] estate - Chapter 6: Basic Views

xafr-odoo committed:
[IMP] estate: Chapter 7
Added property models offer, type and tags

sedi-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,13 @@
+        selection=[('accepted', 'Accepted'), ('refused', 'Refused')],
+        copy=False,)
+    partner_id = fields.Many2one("res.partner", string="Partner", required=True)
+    property_id = fields.Many2one("estate.property", string="Offer", required=True)
```
That's a nasty copy paste (string still indicates it's an Offer 😄). Same for these Many2ones

xafr-odoo replied:
I actually named it `offer_id` at first but then the instructions said to name it `property_id` and it ended up not being displayed anyway.. Should I just omit `string="..."` as much as possible and try to rely on the automated system from odoo you referenced here: https://github.com/odoo/tutorials/pull/740#discussion_r2059848303 ?

sedi-odoo commented on estate/models/estate_property_tags.py:
```diff
@@ -0,0 +1,8 @@
+from odoo import fields, models
+
+
+class EstateTagsModel(models.Model):
```
We try to follow the `_name` for naming a class. This should be `EstatePropertyTag` (keep it singular as well)

xafr-odoo replied:
Yeah I realized afterward that I had pluralized everything. I'll fix it in the next push and also I indeed forgot a word

sedi-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,8 @@
+from odoo import fields, models
+
+
+class EstateTypeModel(models.Model):
```
`EstatePropertyType`

xafr-odoo commented on estate/models/estate_property.py:
```diff
@@ -27,3 +27,8 @@ class EstateModel(models.Model):
         copy=False,
         default='new',
         selection=[('new', 'New'), ('offer_received', 'Offer Received'), ('offer_accepted', 'Offer Accepted'), ('sold', 'Sold'), ('cancelled', 'Cancelled')])
+    property_type_id = fields.Many2one("estate.property.type", string="Property Type")
```
Ok great to know :)

xafr-odoo committed:
[IMP] estate: Chapter 8
Added property computed fields and onchange

xafr-odoo committed:
[IMP] estate: Chapter 9
Added action buttons logic

sedi-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -1,13 +1,26 @@
+    def _inverse_deadline(self):
+        for record in self:
+            if record.create_date:
+                record.validity = (record.date_deadline - fields.Date.to_date(record.create_date)).days
```
Are you sure you need to convert the `record.create_date`? 🤔 
If it is about it being a `datetime` rather than a date, `record.create_date.date()` is cleaner IMO

xafr-odoo replied:
I originally went about it that way because on the Odoo documentation about Date and Datetime fields, the suggested method (and only one that I found to convert from date to datetime) was to use `to_date`...

But yeah this is much cleaner thank you! :)

sedi-odoo reviewed (commented):
Comments about chapter 8

sedi-odoo commented on estate/models/estate_property.py:
```diff
@@ -61,3 +62,19 @@ def _onchange_has_garden(self):
+            if record.state != 'cancelled':
+                record.state = "sold"
+            else:
+                raise UserError("This property is already cancelled and thus cannot be sold anymore.")
```
Nothing bad with this code but, if you use `_()` method it'll be able to handle translations. 

`raise UserError(_("This property is already cancelled and thus cannot be sold anymore."))`

xafr-odoo replied:
Should I get used to adding `_()` for any field that is client facing?

sedi-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -24,3 +24,15 @@ def _inverse_deadline(self):
             if record.create_date:
                 record.validity = (record.date_deadline - fields.Date.to_date(record.create_date)).days
+
+    def action_btn_accept(self):
+        for record in self:
+            record.status = "accepted"
+            record.property_id.buyer_id = self.partner_id.id
+            record.property_id.selling_price = self.price
+        return True
```
When designing such actions you need to think about the usecase. Will you run this method on more than one record at a time? In this case as it is displayed in a list view that is inside a form, it won't.

You can then use a `self.ensure_one()` at the begining to check that it is being run on a single record (not 0 nor more than 1).

If you still want to keep the loop, you should be careful as you're looping over `self` and assigning a value to `record` but yet using `self.partner_id.id` and `self.price`. If you have more than one `partner_id` or different `price` it'll raise an error.

Also, don't forget to update the property state `record.property_id.state = 'offer_accepted'`

xafr-odoo replied:
Oh I see! Thank you for the correction. I'll try and think further ahead

sedi-odoo reviewed (commented):
Chapter 9 comments

xafr-odoo committed:
[IMP] estate: Chapter 10
Constraints

xafr-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -1,13 +1,26 @@
 
 
-class EstateOfferModel(models.Model):
+class EstatePropertyOfferModel(models.Model):
```
Ok will update in the next push :)

xafr-odoo committed:
[IMP] estate: Chapter 11
Add the sprinkles
Fixes according to review comments

xafr-odoo committed:
[IMP] estate: Chapter 12
Inheritance

sedi-odoo commented on .run/rd-demo.run.xml:
```diff
@@ -0,0 +1,22 @@
+<component name="ProjectRunConfigurationManager">
+  <configuration default="false" name="rd-demo" type="Odoo" factoryName="Odoo">
+    <module name="tutorials" />
+    <option name="ENV_FILES" value="" />
+    <option name="INTERPRETER_OPTIONS" value="" />
+    <option name="PARENT_ENVS" value="true" />
+    <envs>
+      <env name="PYTHONUNBUFFERED" value="1" />
+    </envs>
+    <option name="SDK_HOME" value="$PROJECT_DIR$/../../../virtualenvs/18.0/bin/python/" />
+    <option name="WORKING_DIRECTORY" value="" />
+    <option name="IS_MODULE_SDK" value="true" />
+    <option name="ADD_CONTENT_ROOTS" value="true" />
+    <option name="ADD_SOURCE_ROOTS" value="true" />
+    <EXTENSION ID="PythonCoverageRunConfigurationExtension" runner="coverage.py" />
+    <option name="SCRIPT_NAME" value="$PROJECT_DIR$/../../../worktrees/saas-18.3/odoo/odoo-bin" />
+    <option name="PARAMETERS" value="-c $PROJECT_DIR$/.odev/training_odoo.conf -u awesome_owl,awesome_clicker,awesome_gallery,awesome_dashboard,awesome_kanban,estate" />
+    <option name="SHOW_COMMAND_LINE" value="false" />
+    <option name="EMULATE_TERMINAL" value="true" />
+    <method v="2" />
+  </configuration>
+</component>
```
This is not necessary 😄

xafr-odoo replied:
Yeah I understand. I tried to add it to the .gitignore but it did not work and I did not want to unselect it everytime I made a new push.. 😬

sedi-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,50 @@
+            record.status = "refused"
+        return True
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            self.env['estate.property'].browse(vals['property_id']).state = 'offer_received'
+            if vals['price'] < self.env['estate.property'].browse(vals['property_id']).best_offer:
+                raise UserError("Can't make a smaller offer than the best one currently recorded!")
+        return super().create(vals_list)
```
Beware that a browse will run a select in database.

```suggestion
    @api.model_create_multi
    def create(self, vals_list):
        for vals in vals_list:
            estate_property = self.env['estate.property'].browse(vals['property_id'])
            estate_property.state = 'offer_received'
            if vals['price'] < estate_property.best_offer:
                raise UserError("Can't make a smaller offer than the best one currently recorded!")
        return super().create(vals_list)
```

Or even better:


```suggestion
    @api.model_create_multi
    def create(self, vals_list):
        estate_properties = self.env["estate.property"].browse([vals['property_id'] for vals in vals_list if vals['property_id']])
        for vals in vals_list:
            if not vals["property_id"]:
                continue
            estate_property = estate_properties.filtered(lambda ep: ep.id == vals['property_id'])[:1]
            estate_property.state = 'offer_received'
            if vals['price'] < estate_property.best_offer:
                raise UserError("Can't make a smaller offer than the best one currently recorded!")
        return super().create(vals_list)
```

sedi-odoo commented on estate/models/estate_property.py:
```diff
@@ -90,3 +90,8 @@ def _check_offer_price(self):
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_new_or_cancelled_property(self):
+        if any(estate_property.state in ['offer_received', 'offer_accepted', 'sold'] for estate_property in self):
```
This is fine but know that you can do this like this as well:

`if self.filtered(lambda ep: ep.state in ['offer_received', 'offer_accepted', 'sold']):`

or 

`if {'offer_received', 'offer_accepted', 'sold'}.intersect(set(self.mapped('state'))):`

xafr-odoo replied:
Is there a preferred one? If so what is the priority order / where can I find it? :thinking:

sedi-odoo commented on estate/models/res_users.py:
```diff
@@ -0,0 +1,7 @@
+class ResUsers(models.Model):
+    _inherit = "res.users"
+
+    property_ids = fields.One2many("estate.property", "salesman_id", domain="")
```
Don't add `domain` if you don't plan to use it

xafr-odoo replied:
Good catch I was supposed to add a domain there then forgot about it. I will fix that and upload in the next push.

sedi-odoo reviewed (commented):
Review chapter 11 & 12

xafr-odoo committed:
[IMP] estate: Chapter 13
Inter-module interactions

xafr-odoo commented on awesome_gallery/__manifest__.py:
```diff
@@ -4,7 +4,7 @@
         Starting module for "Master the Odoo web framework, chapter 3: Create a Gallery View"
     """,
-
+    'author': "unknown",
```
I just wanted to get rid of the warning messages but yeah should have put odoo. I'm changing it now.

xafr-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -1,11 +1,21 @@
+
+    name = fields.Char('Title', required=True, translate=True)
+    line_ids = fields.One2many("estate.property", "property_type_id")
+    manual_ordering = fields.Integer('Sequence', default=1, help="Used to order stages. Lower is better.")
```
Noted and corrected :)

xafr-odoo committed:
[IMP] estate: Chapter 13
Inter-module interactions

Optimized code to correct a DRY violation that led to unnecessary calls and wasted compute power.

Co-authored-by: Sébastien Dieu <118258604+sedi-odoo@users.noreply.github.com>

xafr-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -1,11 +1,21 @@
+    line_ids = fields.One2many("estate.property", "property_type_id")
+    manual_ordering = fields.Integer('Sequence', default=1, help="Used to order stages. Lower is better.")
+    offer_ids = fields.One2many("estate.property.offer", "property_type_id")
+    offer_count = fields.Integer(compute='_compute_offers')
```
Noted and fixed

sedi-odoo commented on .gitignore:
```diff
@@ -127,3 +127,6 @@ dmypy.json
 .pyre/
+
+# PyCharm Demo Config
+.run/rd-demo.run.xml
```
EOL 😅 
Good thing to add it to the `.gitignore`. But as the file is already tracked, git won't care about this.
You should run `git update-index --skip-worktree .run/rd-demo-run.xml` (can be done on any file you wan't to refresh the tracking on).

(Best way would be to remove the file and then put it back after having removed it with a commit but let's just keep this as is now)

xafr-odoo replied:
Ah I see ok I'll fix it using the last method that's usually how I do it too :)

sedi-odoo commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,35 @@
+    name = fields.Char('Title', required=True, translate=True)
+
+    def action_btn_sold(self):
+        # @SEDI J'ai pas réussi à faire fonctionner avec la plus simple que tu as suggérée. Elle return res.partner(3,) au lieu de res.partner() comme l'autre mais du coup ça marche pas..
+        # partner_id = self.salesman_id.partner_id
+        partner_id = self.env['estate.property'].browse(self.env.context.get('active_id')).salesman_id.partner_id
+        invoicing_price = self.selling_price * 0.06
+
+        values = {
+            'partner_id': partner_id,
```
What error did you get? I think you'd just need to replace `'partner_id': partner_id` by `'partner_id': partner_id.id`

xafr-odoo replied:
Ok that worked thank you :)

sedi-odoo reviewed (commented):
Chapter 13 comments

xafr-odoo committed:
[IMP] estate: Chapter 14
QWeb Templates

xafr-odoo committed:
[IMP] estate: Chapter 15
Coding Guidelines Review

sedi-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -3,8 +3,8 @@
-        <field name="view_mode">list,form</field>
-        <field name="context">{'search_default_availability': True}</field>
+        <field name="view_mode">list,form,kanban</field>
+<!--        <field name="context">{'search_default_availability': True}</field>-->
```
No need to leave a commented line :)

sedi-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -104,4 +104,44 @@
+                            Best Offer:
+                            <field name="best_offer"/>€
+                        </div>
+                        <div  t-if="record.state.raw_value == 'offer_accepted'">
```
You should use a `t-elif` here

xafr-odoo replied:
Ah yes indeed

sedi-odoo commented on estate/models/estate_property.py:
```diff
@@ -57,6 +57,12 @@ def _find_best_offer(self):
+    @api.constrains('expected_price', 'selling_price')
+    def _check_offer_price(self):
+        for record in self:
+            if tools.float_utils.float_compare((record.expected_price * 0.9), record.selling_price, precision_rounding=0.01) >= 0 and not tools.float_utils.float_is_zero(record.selling_price, precision_rounding=0.001):
+                raise ValidationError(_("Offer price cannot be lower than 90% of the expected price."))
```
I'd rather have an import that is 

`from odoo.tools.float_utils impot float_compare, float_is_zero`

That will make the code more readable IMO

sedi-odoo commented on estate/models/estate_property.py:
```diff
@@ -69,6 +75,11 @@ def _onchange_has_garden(self):
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_new_or_cancelled_property(self):
+        if any(estate_property.state in ['offer_received', 'offer_accepted', 'sold'] for estate_property in self):
+            raise UserError("Can't delete a property that is not new of cancelled!")
```
Don't forget the translation 😄

xafr-odoo replied:
Good catch, fixed it! :)

sedi-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -52,3 +40,15 @@ def create(self, vals_list):
+    def action_btn_accept(self):
+        self.ensure_one()
+        self.status = "accepted"
+        self.property_id.buyer_id = self.partner_id.id
+        self.property_id.selling_price = self.price
+        self.property_id.state = 'offer_accepted'
```
This can be done with a `write` and a `dict`

```python
self.property_id.write({
    'buyer_id': self.partner_id.id,
    'selling_price': self.price,
    'state': 'offer_accepted'
})
```

xafr-odoo replied:
Updated! I will try to go about it that way in the future

sedi-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -52,3 +40,15 @@ def create(self, vals_list):
+        self.property_id.state = 'offer_accepted'
+
+    def action_btn_refuse(self):
+        for record in self:
+            record.status = "refused"
```
Can just be done with `records.status = "refused"` no need to loop on them

sedi-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -4,7 +4,7 @@
         <field name="res_model">estate.property</field>
         <field name="view_mode">list,form,kanban</field>
-<!--        <field name="context">{'search_default_availability': True}</field>-->
+        <field name="context">{'search_default_availability': True}</field>
```
Guess that eludes my review on that part 😄

sedi-odoo reviewed (commented):
Chapters 14 & 15

xafr-odoo committed:
[IMP] awesome_owl: Chapter 1
Owl Components

sedi-odoo commented on awesome_owl/static/src/card/card.js:
```diff
@@ -0,0 +1,29 @@
+/** @odoo-module **/
+
+import { Component, useState, markup } from "@odoo/owl";
+import { Counter } from "../counter/counter";
+import {TodoList} from "../todo/todo_list";
+import {TodoItem} from "../todo/todo_item";
```
Be consistent on the format of your imports

sedi-odoo commented on awesome_owl/static/src/card/card.js:
```diff
@@ -0,0 +1,29 @@
+
+    static components = { Counter };
+
+    static template = "awesome_owl.card"
```
You should start component names by a capital letter

`static template = "awesome_owl.Card"`

sedi-odoo commented on awesome_owl/static/src/card/card.xml:
```diff
@@ -0,0 +1,24 @@
+                    <button class="btn btn-primary" t-on-click="toggleCard">Click Me!</button>
+                </h5>
+                <div t-if="this.state.opened">
+                    <t t-slot="default">
+                        <Counter/>
+                    </t>
```
I think the purpose of this part of the tutorial is to create a slot and populate it from outside the component

Basically you should not have the Counter inside of this component but rather in the Playground

sedi-odoo commented on awesome_owl/static/src/card/card.xml:
```diff
@@ -0,0 +1,24 @@
+                        <Counter/>
+                    </t>
+                    <div>
+                        <t t-out="html_bloc_1"/>
+                        <t t-out="html_bloc_2"/>
```
Same about this, tutorial said to use the markup in the playground

sedi-odoo commented on awesome_owl/static/src/counter/counter.js:
```diff
@@ -0,0 +1,25 @@
+        this.state.value++;
+
+        // To avoid errors when the optional onChange is not passed in as attribute
+        if (typeof this.props.onChange !== 'undefined') {
```
Not necessary, a simple check of the prop is sufficient

`if(this.props.onChange) {`

sedi-odoo commented on awesome_owl/static/src/counter/counter.js:
```diff
@@ -0,0 +1,25 @@
+
+export class Counter extends Component {
+    // Inline template here so I can specify it directly in xml, other option is to reference an existing template in the XML file.
+    static template = "awesome_owl.counter"
```
`static template = "awesome_owl.Counter"`

sedi-odoo commented on awesome_owl/static/src/counter/counter.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates xml:space="preserve">
+
+    <t t-name="awesome_owl.counter">
```
`<t t-name="awesome_owl.Counter">`

sedi-odoo commented on awesome_owl/static/src/card/card.xml:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates xml:space="preserve">
+
+    <t t-name="awesome_owl.card">
```
`<t t-name="awesome_owl.Card">`

sedi-odoo commented on awesome_owl/static/src/counter/counter.xml:
```diff
@@ -0,0 +1,10 @@
+
+    <t t-name="awesome_owl.counter">
+        <div>
+            <button class="btn btn-primary" t-on-click="increment">Referenced XML Button</button>
```
Why not `Increment` instead of `Referenced XML Button`?

sedi-odoo commented on awesome_owl/static/src/todo/todo_item.js:
```diff
@@ -0,0 +1,31 @@
+
+export class TodoItem extends Component {
+
+    static template = "awesome_owl.todo_item"
```
`static template = "awesome_owl.TodoItem"`

sedi-odoo commented on awesome_owl/static/src/todo/todo_item.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates xml:space="preserve">
+
+    <t t-name="awesome_owl.todo_item">
```
`<t t-name="awesome_owl.TodoItem">`

sedi-odoo commented on awesome_owl/static/src/todo/todo_item.xml:
```diff
@@ -0,0 +1,13 @@
+
+    <t t-name="awesome_owl.todo_item">
+        <div t-att-class="{'text-muted': props.todo.isCompleted, 'text-decoration-line-through': props.todo.isCompleted}">
+            <input type="checkbox" t-att-checked="props.todo.isCompleted" t-on-change="toggleCheckBox" style="margin-right: 5px"/>
```
That's one way of doing it, know that you could also have done this for the `t-on-change`

`t-on-change="() => props.todo.isCompleted = !props.todo.isCompleted"`

or better, you can call a method on the todo (if you move it to a JS class)

sedi-odoo commented on awesome_owl/static/src/todo/todo_list.js:
```diff
@@ -0,0 +1,48 @@
+    setup() {
+        this.todos = useState([]);
+        this.next_id = 0;
+        this.input = useRef("keyboard_input");
+        onMounted(() => {
+            this.input.el.focus();
+        });
```
Weren't you supposed to use `useAutofocus` instead?

sedi-odoo reviewed (commented):
Chapter 1 JS comments

xafr-odoo committed:
[IMP] awesome_owl: Chapter 2
Build a dashboard

xafr-odoo committed:
[IMP] awesome_owl: Chapter 2
Build a dashboard
